activerecord/lib/active_record/secure_token.rb

```diff
@@ -0,0 +1,38 @@
+module ActiveRecord
+  module SecureToken
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      # Example using has_secure_token
+      #
+      #   # Schema: User(auth_token:string, invitation_token:string)
+      #   class User < ActiveRecord::Base
+      #     has_secure_token
+      #   end
+      #
+      #   user = User.new
+      #   user.save
+      #   user.token # => "44539a6a59835a4ee9d7b112"
+      #   user.regenerate_token # => true
+      def has_secure_token(attribute = :token)
+        # Load securerandom only when has_secure_key is used.
+        require 'securerandom'
+        define_method("regenerate_#{attribute}") { update! attribute => self.class.generate_unique_secure_token(attribute) }
+        before_create { self.send("#{attribute}=", self.class.generate_unique_secure_token(attribute)) }
+      end
+
+      def generate_unique_secure_token(attribute)
+        10.times do |i|
+          SecureRandom.hex(12).tap do |token|
+            if exists?(attribute => token)
```

Comment:
Sorry I missed this earlier... we can probably remove this when @robertomiranda get to switching to base 62. A random UUID (**Universally Unique** Identifier) has 122 random bits. [Wikipedia](http://en.wikipedia.org/wiki/Universally_unique_identifier#Random_UUID_probability_of_duplicates) puts that into perspective:

> To put these numbers into perspective, the annual risk of a given person being hit by a meteorite is estimated to be one chance in 17 billion,[36] which means the probability is about 0.00000000006 (6 × 10−11), equivalent to the odds of creating a few tens of trillions of UUIDs in a year and having one duplicate. In other words, only after generating 1 billion UUIDs every second for the next 100 years, the probability of creating just one duplicate would be about 50%.

With a 24 character base62 sequence, we will have 130 random bits and some change, which makes it over 256 times less likely than a random UUID collision. So basically if a collision is detected, it is much probably more likely that the code or hardware enforcing the uniqueness is at fault than an actual violation, or perhaps a hacker replaced your entropies on your system with anchovies :fish:

Either way, if it happens, you probably have bigger problem to worry about elsewhere :smile:

That said, a uniqueness constraint is still very valuable for protecting against other kind of errors, such as accidentally importing a row from an external source twice, so it should be probably encouraged. What would be nice is that if you can generate this from the generator, i.e. ...

```
rails g model user ... reset_password_token:token
```

...which adds the string column, adds the unique constraint, and inject the `has_secure_token` call in the model. (Same thing for migration generator, scaffold, etc.)

For "plug and play", if you decided not to use a unique constraint, the base62 version should give you the same level of protection... the chance of a collision with the `exists` check *and* a race condition happening at the same time is probably comparable to the chance of collision here (which is, again, practically impossible :smile:)